proof/invariant-abstract/Ipc_AI.thy

```diff
@@ -569,103 +341,342 @@ lemma valid_remove_rights_If[simp]:
+   apply (wp eb hoare_vcg_const_imp_lift hoare_vcg_const_Ball_lift hoare_weak_lift_imp
+           | assumption | simp split del: if_split)+
+      apply (rule cap_insert_assume_null)
+      apply (wp x hoare_vcg_const_Ball_lift cap_insert_cte_wp_at hoare_weak_lift_imp)+
+      apply (rule hoare_vcg_conj_liftE_R')
+       apply (rule derive_cap_is_derived_foo)
+      apply (rule_tac Q' ="\<lambda>cap' s. (vo \<longrightarrow> cap'\<noteq> cap.NullCap \<longrightarrow>
+          cte_wp_at (is_derived (cdt s) (aa, b) cap') (aa, b) s)
+          \<and> (cap'\<noteq> cap.NullCap \<longrightarrow> QM s cap')" for QM
+          in hoare_strengthen_postE_R)
+        prefer 2
+        apply clarsimp
+        apply assumption
+       apply (rule hoare_vcg_conj_liftE_R')
+         apply (rule hoare_vcg_const_imp_liftE_R)
+        apply (rule derive_cap_is_derived)
+      apply (wp derive_cap_is_derived_foo)+
+  apply (clarsimp simp: cte_wp_at_caps_of_state
+                        ex_cte_cap_to_cnode_always_appropriate_strg
+                        real_cte_tcb_valid caps_of_state_valid
+             split del: if_split)
+  apply (clarsimp simp: remove_rights_def caps_of_state_valid
+                        neq_Nil_conv cte_wp_at_caps_of_state
+                        imp_conjR[symmetric] conj_comms
+                 split del: if_split)
+  apply (intro conjI)
+   apply clarsimp
+   apply (case_tac "cap = a",clarsimp)
+   apply (clarsimp simp:masked_as_full_def is_cap_simps)
+   apply (clarsimp simp: cap_master_cap_simps split:if_splits)
+  apply (clarsimp split del: if_split)
+  apply (intro conjI)
+    apply (clarsimp split: if_split)
+   apply (clarsimp)
+  apply (rule ballI)
+  apply (drule(1) bspec)
+  apply clarsimp
+  apply (intro conjI)
+   apply (case_tac "capa = ac",clarsimp+)
+  apply (case_tac "capa = ac")
+   apply (clarsimp simp:masked_as_full_def is_cap_simps split:if_splits)+
+  done
+
+lemmas transfer_caps_loop_pres =
+    transfer_caps_loop_presM[where vo=False and ex=False and em=False, simplified]
+
+lemma transfer_caps_loop_arch[wp]:
+  "\<And>P ep buffer n caps slots mi.
+    \<lbrace>\<lambda>s::'state_ext state. P (arch_state s)\<rbrace>
+      transfer_caps_loop ep buffer n caps slots mi
+    \<lbrace>\<lambda>rv s. P (arch_state s)\<rbrace>"
+  by (rule transfer_caps_loop_pres) wp+
+
+lemma transfer_caps_loop_aobj_at:
+  "arch_obj_pred P' \<Longrightarrow>
+  \<lbrace>\<lambda>s. P (obj_at P' pd s)\<rbrace> transfer_caps_loop ep buffer n caps slots mi \<lbrace>\<lambda>r s::'state_ext state. P (obj_at P' pd s)\<rbrace>"
+  apply (rule hoare_pre)
+   apply (rule transfer_caps_loop_presM[where em=False and ex=False and vo=False, simplified, where P="\<lambda>s. P (obj_at P' pd s)"])
+    apply (wp cap_insert_aobj_at)
+   apply (wpsimp simp: set_extra_badge_def)
+  apply assumption
+  done
+
+end
+
+(* FIXME: can some of these assumptions be proved with lifting lemmas? *)
+locale Ipc_AI_2 = Ipc_AI state_ext_t some_t
```

Comment:
Confused which split you mean. If you mean the one I introduced, it's because of a circular dependency: we need `transfer_caps_loop_presM` and friends inside of Arch in order to instantiate what is now `Ipc_AI_2`, but we can't finish that proof if it uses something that's assumed in `Ipc_AI_2`.

Reply:
If you mean the previous `_cont` split, I have no immediate recollection, but guess that it's also resolving a circular dependency.

Reply:
I meant the one you introduced, thanks for the explanation. That might be worth a small comment to give an idea of the intention for at least one of the locales.

Reply:
I added
```
(* transfer_caps_loop_presM is needed to satisfy assumptions of Ipc_AI_2,
   thus needs to come before Ipc_AI_2 *)
```